test/e2e/main_test.go

```diff
@@ -56,6 +56,20 @@ func init() {
 }
 
 func TestMain(m *testing.M) {
+	os.Exit(testMainWrapper(m))
+}
+
+// testMainWrapper is a wrapper function around the
+// top-level TestMain function and this pattern is
+// needed as os.Exit() doesn't respect any defer calls
+// that may occur during the TestMain workflow. If we
+// we instead doing the heavy-lifting in this function,
+// and then return an integer code that os.Exit can correctly
+// interpret, then the defer call will work.
+//
+// Link to article that explains this behavior:
+// http://blog.englund.nu/golang,/testing/2017/03/12/using-defer-in-testmain.html
```

Comment:
> * http://blog.englund.nu/golang,/testing/2017/03/12/using-defer-in-testmain.html

This is a normal part of the language and [defined in the spec](https://golang.org/pkg/os/#Exit).  I'd link there if you want folks to have official reference.

For reviewers: Because the function never reaches the end of its body the [defer semantics](https://golang.org/ref/spec#Defer_statements) do not apply.  [Simple example](https://play.golang.org/p/_AVhjzsIHgR).  For those with a Java background, think of this like throwing `System.exit()` somewhere in a program.  😄

Reply:
Yeah, that's a good call on linking to the official documentation there. I'll add that now as the tests just flaked and I would need to re-run anyways.